File: app/src/main/java/com/example/instagram/itemfragments/HomeFragment.java
package com.example.instagram.itemfragments;

import android.os.Bundle;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.widget.Toolbar;
import androidx.core.view.MenuProvider;
import androidx.databinding.DataBindingUtil;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.Observer;
import androidx.lifecycle.ViewModelProvider;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Toast;

import com.example.instagram.R;
import com.example.instagram.adapter.ImageAdapter;
import com.example.instagram.adapter.StoryAdapter;
import com.example.instagram.databinding.FragmentHomeBinding;
import com.example.instagram.modal.Photo;
import com.example.instagram.viewmodel.ViewModel;
import com.google.android.material.appbar.AppBarLayout;

import java.util.ArrayList;
import java.util.List;

public class HomeFragment extends Fragment {

    private Toolbar toolbar;
    private RecyclerView postView, storyView;
    private ImageAdapter imageAdapter;
    private ViewModel viewModel;
    private List<Photo> listOfPhoto = new ArrayList<>();
    private FragmentHomeBinding fragmentHomeBinding;

    private StoryAdapter storyAdapter;


    public HomeFragment() {
        // Required empty public constructor
    }


    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        FragmentHomeBinding fragmentHomeBinding = DataBindingUtil.inflate(inflater, R.layout.fragment_home, container, false);
        View view = fragmentHomeBinding.getRoot();
        toolbar = getActivity().findViewById(R.id.toolbar);
        toolbar.setTitle("Instagram");

        viewModel = new ViewModelProvider(requireActivity()).get(ViewModel.class);

        postView = fragmentHomeBinding.postView;
        postView.setLayoutManager(new LinearLayoutManager(getActivity()));

        storyView = fragmentHomeBinding.storyView;
        storyView.setLayoutManager(new LinearLayoutManager(getActivity(), LinearLayoutManager.HORIZONTAL, false));

        imageAdapter = new ImageAdapter(listOfPhoto, getActivity());
        postView.setAdapter(imageAdapter);

        storyAdapter = new StoryAdapter(listOfPhoto, getActivity());
        storyView.setAdapter(storyAdapter);


        viewModel.getViewImages().observe(getActivity(), new Observer<List<Photo>>() {
            @Override
            public void onChanged(List<Photo> photos) {
                listOfPhoto.clear();

                listOfPhoto.addAll(photos);

                imageAdapter.notifyDataSetChanged();
                storyAdapter.notifyDataSetChanged();
            }
        });
        postView.removeOnScrollListener(listener);
        postView.addOnScrollListener(listener);
        return view;
    }

    private RecyclerView.OnScrollListener listener = new RecyclerView.OnScrollListener() {
        @Override
        public void onScrolled(@NonNull RecyclerView recyclerView, int dx, int dy) {
            super.onScrolled(recyclerView, dx, dy);
            Toast.makeText(getActivity(), "scrolled1", Toast.LENGTH_SHORT).show();
            if (dy > 0){
                toolbar.animate().translationY(-(toolbar.getHeight())).setDuration(400).start();
            } else if (dy < 0) {
                toolbar.animate().translationY(0).setDuration(400).start();
            }

        }
    };

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        getActivity().addMenuProvider(menuProvider);
    }


    @Override
    public void onDestroyView() {
        super.onDestroyView();
        getActivity().removeMenuProvider(menuProvider);
    }

    private MenuProvider menuProvider = new MenuProvider() {
        @Override
        public void onCreateMenu(@NonNull Menu menu, @NonNull MenuInflater menuInflater) {
            menuInflater.inflate(R.menu.toolbar_menu, menu);
        }

        @Override
        public boolean onMenuItemSelected(@NonNull MenuItem menuItem) {
            int id = menuItem.getItemId();
            if (id == R.id.like){
                Toast.makeText(getActivity(), "Liked", Toast.LENGTH_SHORT).show();
            } else if (id == R.id.send) {
                Toast.makeText(getActivity(), "Sent", Toast.LENGTH_SHORT).show();
            }
            return true;
        }
    };


}

File: app/src/main/java/com/example/ktinsta/itemfragments/HomeFragment.kt
package com.example.ktinsta.itemfragments

import android.os.Bundle
import android.util.Log
import android.view.*
import android.widget.Toast
import androidx.appcompat.widget.Toolbar
import androidx.core.view.MenuProvider
import androidx.databinding.DataBindingUtil
import androidx.fragment.app.Fragment
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.lifecycleScope
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.example.ktinsta.adapter.ImageAdapter
import com.example.ktinsta.adapter.StoryAdapter
import com.example.ktinsta.dataorexception.DataOrException
import com.example.ktinsta.modal.PexelResponse
import com.example.ktinsta.viewmodel.ImageViewModel
import com.example.instagram.R
import com.example.instagram.databinding.FragmentHomeBinding
import kotlinx.coroutines.launch

class HomeFragment : Fragment() {

    private var binding: FragmentHomeBinding? = null
    private lateinit var toolbar: Toolbar
    private lateinit var postView: RecyclerView
    private lateinit var storyView: RecyclerView
    private lateinit var imageAdapter: ImageAdapter
    private lateinit var storyAdapter: StoryAdapter
    private lateinit var viewModel: ImageViewModel
    private var liveData: LiveData<DataOrException<PexelResponse, Boolean, Exception>> = MutableLiveData()

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        binding = DataBindingUtil.inflate(inflater, R.layout.fragment_home, container, false)
        return binding!!.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        toolbar = requireActivity().findViewById(R.id.toolbar)
        toolbar.title = "Instagram"

        viewModel = ViewModelProvider(requireActivity())[ImageViewModel::class.java]

        postView = binding!!.postView
        postView.layoutManager = LinearLayoutManager(requireActivity())

        storyView = binding!!.storyView
        storyView.layoutManager = LinearLayoutManager(requireActivity(), LinearLayoutManager.HORIZONTAL, false)

        imageAdapter = ImageAdapter(requireActivity())
        postView.adapter = imageAdapter

        storyAdapter = StoryAdapter(requireActivity())

        storyView.adapter = storyAdapter

        viewModel.getPostedImages(6, 10)
        viewModel.data.observe(viewLifecycleOwner) { response ->
            response.data?.let {
                imageAdapter.submitList(it.photos)
                storyAdapter.submitList(it.photos)
            }
        }




        postView.removeOnScrollListener(listener)
        postView.addOnScrollListener(listener)

        requireActivity().addMenuProvider(menuProvider, viewLifecycleOwner)
    }

    private val listener = object : RecyclerView.OnScrollListener() {
        override fun onScrolled(recyclerView: RecyclerView, dx: Int, dy: Int) {
            super.onScrolled(recyclerView, dx, dy)
            if (dy > 0) {
                toolbar.animate().translationY(-toolbar.height.toFloat()).setDuration(400).start()
            } else if (dy < 0) {
                toolbar.animate().translationY(0f).setDuration(400).start()
            }
        }
    }

    private val menuProvider = object : MenuProvider {
        override fun onCreateMenu(menu: Menu, menuInflater: MenuInflater) {
            menuInflater.inflate(R.menu.toolbar_menu, menu)
        }

        override fun onMenuItemSelected(menuItem: MenuItem): Boolean {
            return when (menuItem.itemId) {
                R.id.like -> {
                    Toast.makeText(requireActivity(), "Liked", Toast.LENGTH_SHORT).show()
                    true
                }
                R.id.send -> {
                    Toast.makeText(requireActivity(), "Sent", Toast.LENGTH_SHORT).show()
                    true
                }
                else -> false
            }
        }
    }

    override fun onDestroyView() {
        super.onDestroyView()
        binding = null
    }
}
